File: graphs (undirected + directed)/DFS.java
package graphs;

public class DFS extends GraphTraversal{

	DFS(int vertices) {
		super(vertices);
	}

	public static void main(String[] args) {
		int vertices = 6;
		Graph graph = new Graph(vertices, false);

		graph.insertEdge(1, 2);
		graph.insertEdge(1, 5);
		graph.insertEdge(1, 6);

		graph.insertEdge(2, 5);
		graph.insertEdge(2, 3);

		graph.insertEdge(3, 4);
		graph.insertEdge(4, 5);

		DFS dfs = new DFS(vertices);
		System.out.println("Depth First Search of the graph");
		dfs.depthFirstSearchRecursive(graph, 1);
		
		System.out.println();
		dfs.displayNodeTime();
	}
	
	@Override
	public void processVertexEarly(int vertex) {
		System.out.print(vertex + " ");
	}
	
	@Override
	public void processEdge(int vertex, int nextVertex){

	}
	
	@Override
	public void processVertexLate(int vertex){
	}
	
	public void displayNodeTime() {
		System.out.println();
		for (int i = 1; i < entryTime.length; i++) {
			System.out.println("Node time " + i + ": " + entryTime[i] + "-" + exitTime[i]);
		}
	}

}

File: algorithms/graphs/DFS.mjs
export default class DepthFirstSearch {
  time = 0;
  finished = false;

  constructor(graph) {
    this.graph = graph;
    this.parents = new Array(graph.vertices + 1);
    this.discovered = new Array(graph.vertices + 1);
    this.processed = new Array(graph.vertices + 1);
    this.entryTime = new Array(graph.vertices + 1);
    this.exitTime = new Array(graph.vertices + 1);
  }

  search(v) {
    if (this.finished) return;

    this.discovered[v] = true;
    this.time = this.time + 1;
    this.entryTime[v] = this.time;

    this.processVertexEarly?.(v);

    let node = this.graph.adjacencyList[v];
    while (node) {
      if (!this.discovered[node.y]) {
        this.parents[node.y] = v;
        this.processEdge?.(v, node.y);
        this.search(node.y);
      } else if ((!this.processed[node.y] && this.parents[v] !== node.y) || this.graph.isDirected) {
        this.processEdge?.(v, node.y);
      }
      if (this.finished) return;
      node = node.next;
    }

    this.processVertextLate?.(v);

    this.time = this.time + 1;
    this.exitTime[v] = this.time;
    this.processed[v] = true;
  }
}